src/serac/physics/tests/parameterized_thermomechanics_example.cpp

```diff
@@ -0,0 +1,244 @@
+  FiniteElementState temperature(StateManager::newState(
+      FiniteElementState::Options{.order = p, .coll = std::move(temperature_fec), .name = "theta"}));
+  temperature = theta_ref;
+  simulation.setParameter(temperature, 0);
+
+  double             alpha0    = 1.0e-3;
+  auto               alpha_fec = std::unique_ptr<mfem::FiniteElementCollection>(new mfem::H1_FECollection(p, dim));
+  FiniteElementState alpha(
+      StateManager::newState(FiniteElementState::Options{.order = p, .coll = std::move(alpha_fec), .name = "alpha"}));
+  alpha = alpha0;
+  simulation.setParameter(alpha, 1);
+
+  // set up essential boundary conditions
+  std::set<int> x_equals_0 = {4};
+  std::set<int> y_equals_0 = {2};
+  std::set<int> z_equals_0 = {1};
+
+  auto zero_scalar = [](const mfem::Vector&) -> double { return 0.0; };
+  simulation.setDisplacementBCs(x_equals_0, zero_scalar, 0);
+  simulation.setDisplacementBCs(y_equals_0, zero_scalar, 1);
+  simulation.setDisplacementBCs(z_equals_0, zero_scalar, 2);
+
+  // set up initial conditions
+  auto zero_vector = [](const mfem::Vector&, mfem::Vector& u) -> void { u = 0.0; };
+  simulation.setDisplacement(zero_vector);
+
+  // Finalize the data structures
+  simulation.completeSetup();
+
+  simulation.outputState("paraview");
+
+  // Perform the quasi-static solve
+  double dt   = 1.0;
+  temperature = theta_ref + 1.0;
+  simulation.advanceTimestep(dt);
+
+  // define quantities of interest
+  auto& mesh = serac::StateManager::mesh();
+
+  Functional<double(H1<p, dim>)> qoi({&simulation.displacement().space()});
+  qoi.AddSurfaceIntegral(
+      [=](auto x, auto n, auto displacement) {
+        auto [u, du_dxi] = displacement;
+        return dot(u, n) * ((x[2] > 0.99 * height) ? 1.0 : 0.0);
+      },
+      mesh);
+
+  double initial_qoi = qoi(simulation.displacement());
+  std::cout << "vertical displacement integrated over the top surface: " << initial_qoi << std::endl;
+
+  Functional<double(H1<p, dim>)> area({&simulation.displacement().space()});
+  area.AddSurfaceIntegral([=](auto x, auto /*n*/, auto /*u*/) { return (x[2] > 0.99 * height) ? 1.0 : 0.0; }, mesh);
+
+  std::cout << "total area of the top surface: " << area(simulation.displacement()) << std::endl;
+
+  double exact_area = M_PI_4 * ((outer_radius * outer_radius) - (inner_radius * inner_radius));
+
+  std::cout << "exact area of the top surface: " << exact_area << std::endl;
+
+  std::cout << "average vertical displacement: " << qoi(simulation.displacement()) / area(simulation.displacement())
+            << std::endl;
+
+  double deltaT = 1.0;
+  std::cout << "expected average vertical displacement: " << alpha0 * deltaT * height << std::endl;
+
+  serac::FiniteElementDual adjoint_load(mesh, simulation.displacement().space(), "adjoint_load");
+  auto                     dqoi_du = get<1>(qoi(DifferentiateWRT<0>{}, simulation.displacement()));
```

Comment:
I know `VALUE` and `DERIVATIVE` live in the solid module, but maybe they should be somewhere more accessible so we could use them here?

Reply:
Additionally, using defined `constexpr` values for the parameter indices might make this more readable.

Reply:
`VALUE` and `DERIVATIVE` describe the meaning of the entries of the argument tuples that get passed to the q-function, whereas DifferentiateWRT is telling Functional which `FiniteElementState` to differentiate. 

We could have some constexpr integers that are used to describe the argument indices., but they'll be specific to an individual QoI, as each QoI Functional might depend on different things (or the same set of things, but in a different order).